ocaml/xapi/vgpuops.ml

```diff
@@ -55,10 +55,48 @@ let fail_creation vm vgpu =
         Ref.string_of vgpu.gpu_group_ref
       ]))
 
-let allocate_vgpu_to_gpu ~__context vm host vgpu =
+let allocate_vgpu_to_gpu ?(dry_run=false) ?(pre_allocate_list=[]) ~__context vm host vgpu =
+  (* Get all pGPU from the host *)
   let available_pgpus = Db.Host.get_PGPUs ~__context ~self:host in
+  (* Get all pGPU from the required groups *)
   let compatible_pgpus = Db.GPU_group.get_PGPUs ~__context ~self:vgpu.gpu_group_ref in
   let pgpus = List.intersect compatible_pgpus available_pgpus in
+
+  let pgpu_can_hold_vgpu pgpu vgpu =
+    try Xapi_pgpu.assert_can_run_VGPU ~__context ~self:pgpu ~vgpu;true
+    with e -> false in
+  (* Get all pGPU that can hold the vGPU *)
+  let active_pgpus = List.filter ( fun pgpu -> pgpu_can_hold_vgpu pgpu vgpu.vgpu_ref ) pgpus in
+
+  let remaining_capacity_for_vgpu_from_pgpu vgpu pgpu =
+    let db_remaining =  Helpers.call_api_functions ~__context
+        (fun rpc session_id ->
+           Client.Client.PGPU.get_remaining_capacity ~rpc ~session_id
+             ~self:pgpu ~vgpu_type:vgpu.type_ref) in
+
+    let size_of_vgpu vgpu_ref =
+        let vgpu_type = Db.VGPU.get_type ~__context ~self:vgpu_ref in
+        Int64.to_float (Db.VGPU_type.get_size ~__context ~self:vgpu_type) in
+
+    let current_vgpu_size = size_of_vgpu vgpu.vgpu_ref in
+
+    let size_over_current_vGPU other =
+      let other_size = size_of_vgpu other in
+      other_size /. current_vgpu_size in
+
+    (* Check if any pre_allocation existed, the pre_allocation is a set of vGPU allocation that
```

Comment:
Rather than doing all this pre-allocation checking here, isn't it better to make `Xapi_pgpu.assert_can_run_VGPU` (and functions called from it) aware of the pre-allocations? E.g. it's already taking into account resident and scheduled vGPUs through `Xapi_pgpu_helpers.get_allocated_VGPUs`. It seems unnecessary and confusing to again do computations about sizes and capacity right here.